### PR TITLE
Assignments and labs for lesson 4

### PR DESCRIPTION
**Pull request by cosmonautz**

### CODE REVIEW

File: students/KevinWaldron/lesson04/dict_lab.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+print(s4)
+print(s3.issubset(s2))
+print(s4.issubset(s2))
+
+#Sets 2
+s = set('python')
+s.add('i')
+s2 = frozenset('marathon')
+print(s2.union(s))
+print(s2.intersection(s))
```
Comment (charlesrobison) [inline students/KevinWaldron/lesson04/dict_lab.py]: Looks good!

File: students/KevinWaldron/lesson04/file_lab.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+src = input("Enter file to copy: ")
+dst = input("Enter copy location: ")
+
+with open(src, 'rb') as in_file, open(dst, 'wb') as out_file:
+    while(in_file):
+        line = in_file.read(4096)
+        if not line:
+            break
+        out_file.write(line)
```
Comment (charlesrobison) [inline students/KevinWaldron/lesson04/file_lab.py]: Great

File: students/KevinWaldron/lesson04/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    print("-" * (len(header) - 1))
+    # pack the donors into a list of lists for soting
+    donors_list = []
+    for name in donors:
+        donor_copy = [name]
+        donor_copy.extend(donors[name])
+        donors_list.append(donor_copy)
+    donors_list.sort(key=total_donations, reverse=True)
+    for donor in donors_list:
+        print("{:<20}| ${:>10.2f} | {:>9} | ${:>11.2f}".format(donor[0], donor[1], len(donor) - 2, donor[1]/(len(donor)-2)))
```
Comment (charlesrobison) [inline students/KevinWaldron/lesson04/mailroom.py]: Something going on here with the average gift - not calculating correctly.

File: students/KevinWaldron/lesson04/trigrams.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    if len(sys.argv) != 3:
+        print("Usage: trigrams.py <input-file> <output-file>")
+        exit()
+    in_file = sys.argv[1]
+    out_file = sys.argv[2]
+
+    text = read_text(in_file)
+    trigrams = create_trigrams(text)
+    updated_text = create_text(trigrams)
+    write_text(out_file, updated_text)
```
Comment (charlesrobison) [inline students/KevinWaldron/lesson04/trigrams.py]: Works great!

File: students/KevinWaldron/lesson04/mailroom.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+def donor_entry():
+    """ Handles donor entry menu selection """
+    name = input("\nEnter donor's full name or 'q' to return to the previous menu: ")
+    if(name == "q"):
+        return
+    amount = input("Enter donation amount or 'q' to return to the main menu: ")
+    if (amount == "q"):
+        return
+    if(not name in donors):
```
Comment (natasha-aleksandrova) [inline students/KevinWaldron/lesson04/mailroom.py]: more pythonic -> `if name not in donors`

also no need for parens in if statements.

File: students/KevinWaldron/lesson04/mailroom.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            f.write(create_thank_you({"name":donor, "amount":donors[donor][0]}))
+
+def create_report():
+    """ Handles the create report menu selection """
+    header = "\n{:<20}| Total Given | Num Gifts | Average Gift".format("Donor Name")
+    print(header)
+    print("-" * (len(header) - 1))
+    # pack the donors into a list of lists for soting
+    donors_list = []
+    for name in donors:
```
Comment (natasha-aleksandrova) [inline students/KevinWaldron/lesson04/mailroom.py]: more pythonic to loop over keys and values in dictionary if you are using both:
`for name, donations in donors.items()`

File: students/KevinWaldron/lesson04/mailroom.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        donor_copy = [name]
+        donor_copy.extend(donors[name])
+        donors_list.append(donor_copy)
```
Comment (natasha-aleksandrova) [inline students/KevinWaldron/lesson04/mailroom.py]: can be simplified to:
`donors_list.append([name] + donations)`

File: students/KevinWaldron/lesson04/trigrams.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import random
+
+def read_text(file):
+    """
+    Reads the text from the given input text file into a list of words that were separated by whitespace
+    :param file: input text file
+    :returns: the list of words
+    """
+    text = []
+    with open(file, 'r') as f:
```
Comment (natasha-aleksandrova) [inline students/KevinWaldron/lesson04/trigrams.py]: avoid using `file` for var name (it is a built-in)

(notice different color syntax highlight)

**Pull request merged**